### PR TITLE
Add command line flags to jlm-opt to use Andersen for memory state encoding

### PR DESCRIPTION
**Pull request by haved**
Builds on top of #298 #299 #300

Adds the flags `-AAAndersenAgnostic` and `-AAAndersenRegionAware` to `jlm-opt`, mirroring the corresponding Steensgaard flags, with the only difference being the implementations of `AliasAnalyis` being used.

Feel free to suggest a nicer command line interface for specifying these things, as this was the most direct and straightforward method I could think of.

### CODE REVIEW

File: jlm/llvm/opt/alias-analyses/Optimization.hpp [hunk 1]
```diff
@@ -1,42 +1,45 @@
 /*
  * Copyright 2021 Nico Reißmann <nico.reissmann@gmail.com>
+ * Copyright 2023 Håvard Krogstie <krogstie.havard@gmail.com>
  * See COPYING for terms of redistribution.
  */
 
 #ifndef JLM_LLVM_OPT_ALIAS_ANALYSES_OPTIMIZATION_HPP
 #define JLM_LLVM_OPT_ALIAS_ANALYSES_OPTIMIZATION_HPP
 
+#include "AliasAnalysis.hpp"
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Optimization.hpp:10]: `#include <>`

File: jlm/llvm/opt/alias-analyses/Optimization.hpp [hunk 2]
```diff
@@ -1,42 +1,45 @@
...
+ * @tparam AliasAnalysisPass the subclass of AliasAnalysis to use
+ * @tparam regionAware if true, the RegionAwareMemoryNodeProvider is used
  *
  * @see Steensgaard
+ * @see Andersen
+ * @see AgnosticMemoryNodeProvider
  * @see RegionAwareMemoryNodeProvider
  */
-class SteensgaardRegionAware final : public optimization
+template<typename AliasAnalysisPass, bool regionAware>
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Optimization.hpp:33]: Why using the bool? Why not handing in the provisioning pass directly, similarly to the alias analysis pass?

File: jlm/tooling/Command.cpp [hunk 3]
```diff
@@ -23,6 +23,7 @@
 #include <sys/stat.h>
 
 #include <filesystem>
+#include <jlm/llvm/opt/alias-analyses/AliasAnalysis.hpp>
```
Comment (phate) [inline jlm/tooling/Command.cpp:26]: Why is this needed here? The only change is the inclusion of the header

File: jlm/tooling/CommandLine.cpp [hunk 4]
```diff
@@ -919,6 +935,14 @@ JlmOptCommandLineParser::ParseCommandLineArguments(int argc, char ** argv)
...
               JlmOptCommandLineOptions::ToCommandLineArgument(aASteensgaardRegionAware),
               "Steensgaard alias analysis with region-aware memory state encoding"),
+          ::clEnumValN(
+              aAAndersenAgnostic,
+              JlmOptCommandLineOptions::ToCommandLineArgument(aAAndersenAgnostic),
+              "Andersen alias analysis with agnostic memory state encoding"),
+          ::clEnumValN(
+              aAAndersenRegionAware,
+              JlmOptCommandLineOptions::ToCommandLineArgument(aAAndersenRegionAware),
+              "Andersen alias analysis with region-aware memory state encoding"),
```
Comment (phate) [inline jlm/tooling/CommandLine.cpp:945]: Could we push this over the Steensgaard flags such that they are sorted alphabetically in the `jlm-opt --help` output?

**Review by phate — CHANGES_REQUESTED** [review]

File: jlm/llvm/opt/alias-analyses/Andersen.cpp [hunk 5]
```diff
@@ -0,0 +1,701 @@
+/*
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.cpp:1]: Should be removed from this PR.

File: jlm/llvm/Makefile.sub [hunk 6]
```diff
@@ -41,6 +41,7 @@ LIBLLVM_SRC = \
     jlm/llvm/ir/variable.cpp \
     \
     jlm/llvm/opt/alias-analyses/AgnosticMemoryNodeProvider.cpp \
+    jlm/llvm/opt/alias-analyses/Andersen.cpp \
```
Comment (phate) [inline jlm/llvm/Makefile.sub:44]: Should be removed from this PR.

File: jlm/llvm/opt/alias-analyses/Andersen.hpp [hunk 7]
```diff
@@ -0,0 +1,155 @@
+/*
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Andersen.hpp:1]: Should be removed from this PR.

File: jlm/llvm/opt/alias-analyses/Optimization.hpp [hunk 8]
```diff
@@ -1,42 +1,46 @@
...
 
+#include <type_traits>
+
 namespace jlm::llvm::aa
 {
 
-/** \brief Steensgaard alias analysis with agnostic memory state encoding
+/** Applies alias analysis and memory state encoding.
+ * Uses the information collected during alias analysis and
+ * the memory nodes provided by the memory node provided
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Optimization.hpp:20]: "and the memory nodes provided by the memory node providerR"

File: jlm/llvm/opt/alias-analyses/Optimization.hpp [hunk 9]
```diff
@@ -1,42 +1,46 @@
...
+ * to reencode memory state edges beteen the operations touching memory.
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Optimization.hpp:21]: "betWeen"

File: jlm/llvm/opt/alias-analyses/Optimization.hpp [hunk 10]
```diff
@@ -1,42 +1,46 @@
...
+ * @tparam MemoryNodeProviderPass the subclass of MemoryNodeProvider to use
  *
  * @see Steensgaard
+ * @see Andersen
+ * @see AgnosticMemoryNodeProvider
  * @see RegionAwareMemoryNodeProvider
  */
-class SteensgaardRegionAware final : public optimization
+template<typename AliasAnalysisPass, typename MemoryNodeProviderPass>
+class MemoryStateEncodingPass final : public optimization
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/Optimization.hpp:34]: I am not fully sold on the name "MemoryStateEncodingPass". I am somehow bothered by the suffix "Pass" here, but I have also a hard time coming up with something better. Maybe AliasAnalysisStateEncoder?

Comment (phate) [reply jlm/llvm/opt/alias-analyses/Optimization.hpp:34]: Leave it if you do not like AliasAnalysisStateEncoder and nothing else comes to your mind.

File: jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp [hunk 11]
```diff
@@ -19,20 +19,20 @@ PointerObjectSet::AddPointerObject(PointerObjectKind kind)
   return PointerObjects_.size() - 1;
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp:19]: Should be removed from this PR.

File: jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp [hunk 12]
```diff
@@ -190,7 +190,7 @@ class PointerObjectSet final
   /**
    * Internal helper function for adding PointerObjects, use the Create* methods instead
    */
-  PointerObject::Index
+  [[nodiscard]] PointerObject::Index
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:193]: Should be removed from this PR.

File: tests/jlm/llvm/opt/alias-analyses/Makefile.sub [hunk 13]
```diff
@@ -1,5 +1,6 @@
 TESTS += \
 	jlm/llvm/opt/alias-analyses/TestAgnosticMemoryNodeProvider \
+	jlm/llvm/opt/alias-analyses/TestAndersen \
```
Comment (phate) [inline tests/jlm/llvm/opt/alias-analyses/Makefile.sub:3]: Should be removed from this PR.

File: tests/jlm/llvm/opt/alias-analyses/TestAndersen.cpp [hunk 14]
```diff
@@ -0,0 +1,906 @@
+/*
```
Comment (phate) [inline tests/jlm/llvm/opt/alias-analyses/TestAndersen.cpp:1]: Should be removed from this PR.

**Review by phate — CHANGES_REQUESTED** [review]
Mostly cleaning up after the Andersen alias analysis got merged.

**Comment by haved** [comment]
Merging with both the old branch and master resolved all the extra changes not part of this PR.
I used the name you suggested, I liked it very well, and was not really happy with the old one.

I also moved AAAndersen above AASteensgaard in all the places they are listed, for consistency.

**Review by phate — APPROVED** [review]
Solid work. Thanks a lot!

**Pull request merged**